### PR TITLE
*: support require-secure-transport startup option

### LINKED ISSUE #15306: Add require_secure_transport
## Feature Request

**Is your feature request related to a problem? Please describe:**

We should add support for a require_secure_transport startup option that matches the behavior of MySQL Server:

https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_require_secure_transport

https://dev.mysql.com/doc/refman/5.7/en/using-encrypted-connections.html#mandatory-encrypted-connections

**Describe the feature you'd like:**

"With require_secure_transport enabled, client connections to the server are required to use some form of secure transport, and the server permits only TCP/IP connections that use SSL, or connections that use a socket file (on Unix) or shared memory (on Windows). The server rejects nonsecure connection attempts, which fail with an ER_SECURE_TRANSPORT_REQUIRED error."

**Describe alternatives you've considered:**

Without a require_secure_transport setting, an administrator must be careful to create all users with a REQUIRE TLS/ISSUE/CIPHER/etc. clause.

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 
 
fixes #15306

### What is changed and how it works?

- add `--require-secure-transport` startup option
- reject non-TLS client connect to TiDB if `require-secure-transport` be enabled 
- startup failure if `require-secure-transport` be enabled but no avaliable cert/key/ca
- reload failure if `require-secure-transport` be enabled event if with `no rollback on error`

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test 
```
enable `require-secure-transport` and try non-TLS client
```

Code changes

 - imple change

Side effects

 - n/a

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository

Release note

 - support --require-secure-transport startup option.

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/15341)

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -209,12 +209,13 @@ func NewServer(cfg *config.Config, driver IDriver) (*Server, error) {
 	tlsConfig, err := util.LoadTLSCertificates(s.cfg.Security.SSLCA, s.cfg.Security.SSLKey, s.cfg.Security.SSLCert)
 	if err != nil {
 		logutil.BgLogger().Error("secure connection cert/key/ca load fail", zap.Error(err))
-		return nil, err
```
Comment (jackysp) [inline server/server.go]: why remove this line?

File: tidb-server/main.go [hunk 2]
```diff
@@ -68,33 +68,34 @@ import (
...
+	nmStatusPort             = "status"
+	nmMetricsAddr            = "metrics-addr"
+	nmMetricsInterval        = "metrics-interval"
+	nmDdlLease               = "lease"
+	nmTokenLimit             = "token-limit"
+	nmPluginDir              = "plugin-dir"
+	nmPluginLoad             = "plugin-load"
+	nmRepairMode             = "repair-mode"
+	nmRepairList             = "repair-list"
+	nmRequireSecureTransport = "require-secure-transport"
```
Comment (jackysp) [inline tidb-server/main.go]: I know it is the same name as MySQL's, but I'm not sure adding it is approved.
\cc @siddontang

File: server/conn.go [hunk 3]
```diff
@@ -521,6 +522,8 @@ func (cc *clientConn) readOptionalSSLRequestAndHandshakeResponse(ctx context.Con
 				return err
 			}
 		}
+	} else if config.GetGlobalConfig().Security.RequireSecureTransport {
+		return errors.New("secure transport is required")
```
Comment (kolbe) [inline server/conn.go]: I think we should add the same error that is used in MySQL: https://dev.mysql.com/doc/refman/8.0/en/server-error-reference.html#error_er_secure_transport_required

File: server/server.go (hunk 1 above)
Comment (lysu) [reply server/server.go]: after add require-secure-connection option, it's better follow mysql logic, server will stop if require-secure-connection = true, but we should not stop if user didn't want to require tls.

File: tidb-server/main.go (hunk 2 above)
Comment (siddontang) [reply tidb-server/main.go]: PTAL @tennix @iamxy

Comment (siddontang) [reply tidb-server/main.go]: I use `mysql --help` but don't find hat it uses this in command flags.

Comment (jackysp) [reply tidb-server/main.go]: It is a flag of `mysqld`, try `mysqld --verbose --help`

**Comment by shenli** [comment]
@kolbe @jackysp PTAL

**Pull request merged**